bin/i18n/i18n_script_utils.rb

```diff
@@ -135,30 +135,47 @@ def self.get_level_url_key(script, level)
   def self.get_level_from_url(url)
     # memoize to reduce repeated database interactions
     @levels_by_url ||= Hash.new do |hash, new_url|
-      url_regex = %r{https://studio.code.org/s/(?<script_name>[A-Za-z0-9\s\-_]+)/stage/(?<stage_pos>[0-9]+)/(?<level_info>.+)}
-      matches = new_url.match(url_regex)
-
-      hash[new_url] =
-        if matches.nil?
-          project_url_regex = %r{https://studio.code.org/p/(?<project_name>[A-Za-z0-9\s\-_]+)}
-          project_matches = new_url.match(project_url_regex)
-          if project_matches.nil?
-            STDERR.puts "could not find level for url: #{new_url}"
-            nil
+      route_params = Rails.application.routes.recognize_path(new_url)
+
+      level =
+        case route_params[:controller]
+        when "projects"
+          Level.find_by_name(ProjectsController::STANDALONE_PROJECTS[params[:key]][:name])
+        when "script_levels"
+          script = Script.get_from_cache(route_params[:script_id])
```

Comment:
Might be nice for readability to break out the logic in each of these cases into their own functions, YMMV though.